geth/peers/peerpool_test.go

```diff
@@ -99,34 +102,37 @@ func (s *PeerPoolSimulationSuite) getPoolEvent(events <-chan string) string {
 	case ev := <-events:
 		return ev
 	case <-time.After(time.Second):
-		s.FailNow("timed out waiting for a peer")
+		s.FailNow("timed out waiting a pool event")
 		return ""
 	}
 }
 
 func (s *PeerPoolSimulationSuite) TestSingleTopicDiscoveryWithFailover() {
-	poolEvents := make(chan string, 1)
-	summaries := make(chan []*p2p.PeerInfo, 1)
+	var err error
+
+	poolEvents := make(chan string)
+	summaries := make(chan []*p2p.PeerInfo)
 	signal.SetDefaultNodeNotificationHandler(func(jsonEvent string) {
 		var envelope struct {
 			Type  string
 			Event json.RawMessage
 		}
 		s.NoError(json.Unmarshal([]byte(jsonEvent), &envelope))
-		switch envelope.Type {
-		case signal.EventDiscoveryStarted:
-			poolEvents <- envelope.Type
-		case signal.EventDiscoveryStopped:
-			poolEvents <- envelope.Type
-		case signal.EventDiscoverySummary:
-			poolEvents <- envelope.Type
-			var summary []*p2p.PeerInfo
-			s.NoError(json.Unmarshal(envelope.Event, &summary))
-			summaries <- summary
-		}
 
+		go func() {
```

Comment:
why move to goroutine?

Reply:
Because I was getting timeouts due to deadlock instead of a meaningful error which condition failed.

Reply:
However, the issue with goroutine is that it introduces uncertainty in the order of events... So, actually buffered channels have to be used.